### PR TITLE
ARROW-16675: [C++] Wrong Tell() result from BufferedOutputStream in an edge case 

### PR DESCRIPTION
**Pull request by zhztheplayer**

### CODE REVIEW

File: cpp/src/arrow/io/buffered_test.cc [hunk 1]
```diff
@@ -297,6 +297,19 @@ TEST_F(TestBufferedOutputStream, Tell) {
   AssertTell(0);
 }
 
+TEST_F(TestBufferedOutputStream, Tell2) {
+  OpenBuffered();
+  AssertTell(0);
+
+  const std::string datastr = GenerateRandomData(10000);
```
Comment (cyb70289) [inline cpp/src/arrow/io/buffered_test.cc:304]: A simple test string should be enough. It costs time to generate random data.
I think we can move this case to existing `Tell` tests, not necessary for a new test.

**Review by cyb70289 — COMMENTED** [review]
Thanks for the fix @zhztheplayer !
I have one comments.

File: cpp/src/arrow/io/buffered_test.cc [hunk 2]
```diff
@@ -293,6 +293,14 @@ TEST_F(TestBufferedOutputStream, Tell) {
 
   ASSERT_OK(buffered_->Close());
 
+  // write large buffer after raw_pos is cached
+  OpenBuffered(3, false);
```
Comment (cyb70289) [inline cpp/src/arrow/io/buffered_test.cc:297]: `OpenBuffered(/*buffer_size=*/3);` should make it easier to catch the intent of this test.

File: cpp/src/arrow/io/buffered_test.cc [hunk 3]
```diff
@@ -293,6 +293,14 @@ TEST_F(TestBufferedOutputStream, Tell) {
...
+  AssertTell(0);
+  ASSERT_OK(buffered_->Write(std::string(10, '0').data(), 5));
```
Comment (cyb70289) [inline cpp/src/arrow/io/buffered_test.cc:299]: Looks to me this code is deferencing a dangling pointer. Please define a string variable.

**Review by cyb70289 — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/io/buffered_test.cc (hunk 3 above)
Comment (zhztheplayer) [reply cpp/src/arrow/io/buffered_test.cc]: Oops. Thanks for catching.

File: cpp/src/arrow/io/buffered_test.cc (hunk 2 above)
Comment (zhztheplayer) [reply cpp/src/arrow/io/buffered_test.cc]: done

File: cpp/src/arrow/io/buffered_test.cc [hunk 4]
```diff
@@ -293,6 +293,15 @@ TEST_F(TestBufferedOutputStream, Tell) {
 
   ASSERT_OK(buffered_->Close());
 
+  // write long bytes after raw_pos is cached
+  const std::string basic_string = "1234568790";
+  OpenBuffered(/*buffer_size=*/3, false);
```
Comment (cyb70289) [inline cpp/src/arrow/io/buffered_test.cc:298]: Last nit comment :)
 `false` looks not necessary. If you think it's important, please also add `/*append=*/false`, boolean variables are hard to reason from caller side.

File: cpp/src/arrow/io/buffered_test.cc [hunk 5]
```diff
@@ -293,6 +293,15 @@ TEST_F(TestBufferedOutputStream, Tell) {
...
+  AssertTell(0);
+  ASSERT_OK(buffered_->Write(basic_string.data(), 5));
```
Comment (cyb70289) [inline cpp/src/arrow/io/buffered_test.cc:300]: Is `buffered_->Write("1234567890", 5)` enough for the test?

Comment (zhztheplayer) [reply cpp/src/arrow/io/buffered_test.cc]: done

File: cpp/src/arrow/io/buffered_test.cc (hunk 4 above)
Comment (zhztheplayer) [reply cpp/src/arrow/io/buffered_test.cc]: done

File: cpp/src/arrow/io/buffered.cc [hunk 6]
```diff
@@ -144,6 +144,8 @@ class BufferedOutputStream::Impl : public BufferedBase {
       RETURN_NOT_OK(FlushUnlocked());
       DCHECK_EQ(buffer_pos_, 0);
       if (nbytes >= buffer_size_) {
+        // Invalidate cached raw pos
+        raw_pos_ = -1;
```
Comment (pitrou) [inline cpp/src/arrow/io/buffered.cc:148]: Could this instead be `raw_pos_ += nbytes`?

Comment (pitrou) [reply cpp/src/arrow/io/buffered.cc:148]: Right: in case of error, should probably set it to -1.

**Pull request merged**